### ISSUE TITLE
error re-using consumer tag: channel closed

### ISSUE THREAD

**Issue by derickbailey** [arobson/rabbot#4]
I'm working on updating my Rabbus library to use Rabbot, and running into an issue with my tests.

Generally speaking, the drop-in replacement is working when i run my code. But when I try to run my tests, things go bad fast. 

I can run a single test - any test - and it will work fine. But as soon as I try to run two tests, or 1 jasmine test with two `it` blocks in it, i end up getting a channel closed error:

`Error: Channel ended, no reply will be forthcoming`

I'm looking at my RabbitMQ log files and I see this:

```
=INFO REPORT==== 3-Mar-2016::08:22:33 ===
accepting AMQP connection <0.3764.0> (127.0.0.1:49377 -> 127.0.0.1:5672)

=ERROR REPORT==== 3-Mar-2016::08:22:33 ===
Error on AMQP connection <0.3764.0> (127.0.0.1:49377 -> 127.0.0.1:5672, vhost: '/rabbus-specs', user: 'rabbus', state: running), channel 3:
operation basic.consume caused a connection exception not_allowed: "attempt to reuse consumer tag 'frodo.local.grunt.1454.send-receive.q.1'"

=INFO REPORT==== 3-Mar-2016::08:22:33 ===
closing AMQP connection <0.3764.0> (127.0.0.1:49377 -> 127.0.0.1:5672)
```

it looks like the consumers in my code are attempting to re-use the consumer tag that was generated by rabbot, and this is causing rabbitmq to close things and go into an error state.

i'm still digging into this, and wondering what about jasmine is causing this (if it is jasmine)... but any guidance or direction on when / where / why / how this consumer tag is created would be helpful

**Comment by derickbailey** [comment]
I've reproduced this problem with rabbot, now. The problem occurs when trying to start two consumers on the same queue, within the same process

the code in question is the consumer tag creation: https://github.com/arobson/rabbot/blob/master/src/info.js#L11-L16

you can recreate this problem with this code:

```js
var rabbot = require("rabbot");

var config = {
  connection: { host: "localhost", user: "bob", pass: "whatever", vhost: "test" },
  exchanges: [{ name: "rabbot.ex.1", type: "direct" }],
  queues: [{ name: "rabbot.q.1", subscribe: false }],
  bindings: [{ exchange: "rabbot.ex.1", target: "rabbot.q.1", keys: "some.key" }]
};

rabbot.configure(config)
  .then(function(){
    receive();
    receive();
  })
  .then(undefined, reportError);

function receive(){
  rabbot.handle("some.type", function(msg){
    msg.ack();
  });
  rabbot.startSubscription("rabbot.q.1");
}

function reportError(err){
  console.log(err.stack);
  process.exit();
}

process.on("uncaughtException", reportError);
process.on("unhandledRejection", reportError);
```

note the two calls to the `receive` method, which is what causes this.

this seems like it might be a non-issue off-hand, since you would have to try and subscribe to the same queue twice, from the same process instance, but i would expect this scenario to be handled. 

my jasmine tests are already causing this scenario, and i can see some potential for this happening if someone were using domains or clustering or whatever other high-availability things that might have multiple instances of code in a single process instance

my suggested solution: add a unique chunk to the consumer tag... whether it's sequential, random, uuid, or whatever

**Comment by derickbailey** [comment]
to verify my thoughts on a solution, i tried out a simple workaround for the `createConsumerTag` method:

```js
var consumerCount = 0;

function createConsumerTag( queueName ) {
	consumerCount += 1;
	return format( "%s.%s.%s", consumerId, queueName, consumerCount);
}
```

this code allowed me to call the `receive` method twice, starting two subscriptions to the same queue from the same process

**Comment by arobson** [comment]
@derickbailey - just so I'm clear - you don't have a use case where you necessarily _want_ two subscriptions to the same queue from the same process, but there are circumstances that _could_ lead to that except rabbot blows up.

I'll look into this one soon and let you know when I have a fix in place, I just want to be certain that if I make this behave like a singleton, it won't break something else for your usage.

**Comment by derickbailey** [comment]
> you don't have a use case where you necessarily want two subscriptions to the same queue from the same process

correct - i don't really *want* two subscriptions to the same queue, but that is what happens due to the use of `beforeEach` in a jasmine test with multiple `it` blocks.

so, the use case would be test suites

**Comment by derickbailey** [comment]
> I just want to be certain that if I make this behave like a singleton, it won't break something else for your usage.

i think this should be fine, as long as rabbot is forwarding the existing subscription through all handlers.

so, if i create the subscription with handler1 and then later on try to re-create the subscription and add handler2, i would want the existing subscription to be used for handler2

**Comment by arobson** [comment]
@derickbailey - this should be fixed in 1.0.0-9. If the subscription is already in place, it just returns the promise with the consumer tag, otherwise the subscription is started.

Subscriptions are completely independent of handlers so there's no reason why adding a handler later should require any change to existing subscriptions.
